File: src/nexus_scalp/database/connection_url.py
"""Pure parse / build / redact helpers for PostgreSQL connection URLs.

DATABASE PORTABILITY — connection-string entry (2026-09-23): an operator may
type ONE URL (``postgresql://user@host:port/db?sslmode=require``) instead of
filling six discrete fields.  This module is the server-side authority for
that string; the react console mirrors it by contract
(``frontend/.../connectionUrl.ts``), never by import.

INVARIANTS (contract §3.1):
  * the accepted scheme set is EXACTLY ``{"postgresql", "postgres", "pgsql"}``
    — the same set ``DatabaseProvider.parse`` accepts;
  * the module NEVER raises at the configuration boundary: a bad URL returns
    a ``ParseFailure`` with a human sentence, never an exception traceback
    (exception text may embed credentials or internals);
  * ``build_pg_url`` NEVER writes a password into the emitted URL;
  * ``redact_url`` replaces any password with ``***`` for display/logs;
  * no import of the web layer — this is a pure dependency-free module.

EXTEND: add a new accepted field by teaching ``_apply_query_args`` and
``build_pg_url`` about it, and mirror the change in the client module.
"""

from __future__ import annotations

from typing import TypedDict
from urllib.parse import unquote

from typing_extensions import TypeIs

#: Schemes accepted on both sides of the wire (contract §3.1).
#: Mirrors ``DatabaseProvider.parse``'s postgres-family set exactly.
PG_URL_SCHEMES: frozenset[str] = frozenset({"postgresql", "postgres", "pgsql"})

#: Query-string keys understood by the parser (lower-cased on read).
_SSL_MODE_KEYS: frozenset[str] = frozenset({"sslmode", "ssl_mode"})

#: SQL-standard SSL mode values (used only to normalize case/spelling of a
#: value the operator already typed — never to invent one).
_SSL_MODES: frozenset[str] = frozenset(
    {
        "disable",
        "allow",
        "prefer",
        "require",
        "verify-ca",
        "verify-full",
    }
)


class ParsedPgConfig(TypedDict):
    """Discrete PostgreSQL fields derived from a URL (password EXCLUDED).

    ``password`` is deliberately absent: the route layer routes it to the
    OS SecretStore and this dict is what travels back to the browser.
    """

    host: str
    port: int
    database: str
    username: str
    ssl_mode: str


class ParseFailure(TypedDict):
    """A refusal, as a human sentence (never an exception text)."""

    reason: str


def _sentence(reason: str) -> ParseFailure:
    return {"reason": reason}


def _norm_scheme(scheme: str | None) -> str:
    return (scheme or "").strip().lower()


def _fail_bad_scheme(scheme: str) -> ParseFailure:
    pretty = scheme or "(none)"
    return _sentence(
        f"Connection URL must use the postgresql scheme (postgresql, postgres "
        f"or pgsql); got '{pretty}'."
    )


def _parse_port(raw: str | None) -> int | None:
    """Return the port as int, or None when it is not a usable 1..65535 value."""
    if raw is None or raw == "":
        return None
    text = raw.strip()
    if not text.isdigit():
        return None
    value = int(text)
    if not 1 <= value <= 65535:
        return None
    return value


def _apply_query_args(fields: dict[str, object], query: str) -> None:
    """Fold the query string's sslmode into ``fields`` (in place)."""
    if not query:
        return
    for pair in query.split("&"):
        if not pair:
            continue
        key, _, value = pair.partition("=")
        name = unquote(key).strip().lower()
        if name in _SSL_MODE_KEYS:
            fields["ssl_mode"] = unquote(value).strip()


def _normalize_ssl_mode(value: str) -> str:
    """Canonicalize a typed ssl mode; empty stays empty (never invented)."""
    text = (value or "").strip()
    if not text:
        return ""
    lowered = text.lower()
    return lowered if lowered in _SSL_MODES else text


def parse_pg_url(raw: str) -> ParsedPgConfig | ParseFailure:
    """Parse ``postgresql://user[:***@host:port/db?sslmode=require``.

    Accepts the ``postgres://`` and ``pgsql://`` spellings too.  Rejects:
    a scheme outside the postgres family, a missing host, a non-numeric port,
    a port outside 1..65535 and a URL with no database name in its path.
    Never raises.
    """
    if not isinstance(raw, str) or not raw.strip():
        return _sentence("A connection URL is required.")

    text = raw.strip()
    if "://" not in text:
        return _fail_bad_scheme(text.split(":", 1)[0] if ":" in text else "")

    scheme, rest = text.split("://", 1)
    scheme = _norm_scheme(scheme)
    if scheme not in PG_URL_SCHEMES:
        return _fail_bad_scheme(scheme)

    if not rest:
        return _sentence("Connection URL is missing its host and database name.")

    # Split off the query string before touching credentials/host.
    path_and_query = rest.split("?", 1)
    authority_path = path_and_query[0]
    query = path_and_query[1] if len(path_and_query) > 1 else ""

    # rsplit: an IPv6 host contains ':' but the userinfo '@' is the real seam.
    creds_sep = authority_path.rfind("@")
    if creds_sep >= 0:
        userinfo = authority_path[:creds_sep]
        hostport_path = authority_path[creds_sep + 1 :]
    else:
        userinfo = ""
        hostport_path = authority_path

    username = unquote(userinfo.split(":", 1)[0]).strip() if userinfo else ""

    # netloc may carry a port AND a path: split the first '/'.
    path_sep = hostport_path.find("/")
    if path_sep >= 0:
        hostport = hostport_path[:path_sep]
        path = hostport_path[path_sep:]
    else:
        hostport = hostport_path
        path = ""

    host: str
    port: int
    if hostport.startswith("["):
        # IPv6 literal: [::1]:5432
        close = hostport.find("]")
        if close < 0:
            return _sentence("Connection URL has an unterminated IPv6 host literal.")
        host = hostport[1:close]
        tail = hostport[close + 1 :]
        port_value = _parse_port(tail.lstrip(":") if tail else None)
        if tail and port_value is None:
            return _sentence("Connection URL port must be a number between 1 and 65535.")
        port = port_value if port_value is not None else 5432
    elif ":" in hostport:
        host_text, _, port_text = hostport.partition(":")
        host = host_text.strip()
        port_value = _parse_port(port_text)
        if port_value is None:
            return _sentence("Connection URL port must be a number between 1 and 65535.")
        port = port_value
    else:
        host = hostport.strip()
        port = 5432

    if not host:
        return _sentence("Connection URL is missing its host.")

    database = unquote(path.lstrip("/")).strip()
    if not database:
        return _sentence("Connection URL is missing its database name.")

    fields: dict[str, object] = {
        "host": host,
        "port": port,
        "database": database,
        "username": username,
        "ssl_mode": "",
    }
    _apply_query_args(fields, query)
    fields["ssl_mode"] = _normalize_ssl_mode(str(fields["ssl_mode"]))

    return fields  # type: ignore[return-value]


def is_parse_failure(result: ParsedPgConfig | ParseFailure) -> TypeIs[ParseFailure]:
    """Discriminate the parse_pg_url union for type checkers.

    mypy does not narrow a TypedDict union from a literal `in` test, so the
    routes layer guards with this TypeGuard instead of `"reason" in parsed`.
    """
    return "reason" in result


def build_pg_url(cfg: ParsedPgConfig) -> str:
    """Inverse of :func:`parse_pg_url`: fields -> a display URL.

    The emitted URL NEVER carries a password: it is a round-trip of the
    discrete fields only, safe to show in the UI and safe to log.
    """
    host = str(cfg.get("host") or "")
    host_part = host
    if ":" in host_part and not host_part.startswith("["):
        host_part = f"[{host_part}]"
    username = str(cfg.get("username") or "")
    user_part = f"{username}@" if username else ""
    port = cfg.get("port") or 5432
    database = str(cfg.get("database") or "")
    ssl_mode = _normalize_ssl_mode(str(cfg.get("ssl_mode") or ""))
    url = f"postgresql://{user_part}{host_part}:{port}/{database}"
    if ssl_mode:
        url = f"{url}?sslmode={ssl_mode}"
    return url


def redact_url(raw: str) -> str:
    """Mask any password in a URL for display: ``user:***@host``.

    Used for the UI echo of what the operator typed.  A URL with no
    userinfo or no password separator is returned unchanged.
    """
    if not isinstance(raw, str) or "://" not in raw:
        return raw
    scheme, rest = raw.split("://", 1)
    if "@" not in rest:
        return raw
    creds, tail = rest.rsplit("@", 1)
    if ":" not in creds:
        return raw
    user, _pw = creds.split(":", 1)
    return f"{scheme}://{user}:***@{tail}"

File: frontend/src/features/database/ui/connectionUrl.ts
/**
 * Pure client-side mirror of the server connection-URL helpers (contract §3.1).
 *
 * OWNER: lane B (db-provider-pro) — this lane owns future edits.
 * CONSUMES: nothing at runtime (pure, no React, no fetch); the server authority
 *   is `src/nexus_scalp/database/connection_url.py` — mirrored BY CONTRACT,
 *   never imported (cross-lane dependency rule, contract §2.1).
 * PROVIDES: `parsePgUrl` / `buildPgUrl` / `maskUrl` + the `ParsedPgConfig` and
 *   `UrlParseFailure` shapes the connection panel and the URL field consume.
 * INVARIANTS:
 *   - the accepted scheme set is EXACTLY {postgresql, postgres, pgsql} — the
 *     same set `DatabaseProvider.parse` accepts;
 *   - NEVER throws at the UI boundary: a malformed URL returns a human
 *     sentence (`UrlParseFailure.reason`), so a bad string never reaches the
 *     backend and an exception text (which may embed credentials) is never
 *     surfaced;
 *   - `buildPgUrl` NEVER writes a password into the emitted URL — it is a
 *     round-trip of the discrete fields only, safe to display and to log;
 *   - `maskUrl` replaces any password with `***` (display echo only).
 *   - This module is unit-tested by `tests/js/test_connection_url.mjs`, so it
 *     must stay free of non-erased imports (node runs it directly with types
 *     stripped; no `@/` alias may resolve at runtime).
 * EXTEND: add a new parsed field by teaching `applyQueryArgs` and `buildPgUrl`
 *   about it, and mirror the change on the server side.
 */

export type ParsedPgConfig = {
  host: string;
  port: number;
  database: string;
  username: string;
  ssl_mode: string;
  /** Present ONLY when the URL carried it — the server-side echo excludes it
   *  (contract §3.2); the client keeps it transiently to populate the form's
   *  password field, and `buildPgUrl` never writes it back. */
  password?: string;
};

export type UrlParseFailure = {
  /** A human sentence; never an exception/traceback text. */
  reason: string;
};

/**
 * Schemes accepted on both sides of the wire (contract §3.1).
 * Mirrors `DatabaseProvider.parse`'s postgres-family set exactly.
 */
export const PG_URL_SCHEMES: ReadonlySet<string> = new Set(["postgresql", "postgres", "pgsql"]);

/** Query-string keys understood by the parser (lower-cased on read). */
const SSL_MODE_KEYS: ReadonlySet<string> = new Set(["sslmode", "ssl_mode"]);

/**
 * SQL-standard SSL mode values — used ONLY to normalize the case of a value
 * the operator already typed, never to invent one.
 */
const SSL_MODES: ReadonlySet<string> = new Set([
  "disable",
  "allow",
  "prefer",
  "require",
  "verify-ca",
  "verify-full",
]);

const DEFAULT_PG_PORT = 5432;

/** True when `parsePgUrl` refuses `raw` (never throws). */
export function isUrlParseFailure(v: ParsedPgConfig | UrlParseFailure): v is UrlParseFailure {
  return typeof (v as UrlParseFailure).reason === "string";
}

function fail(reason: string): UrlParseFailure {
  return { reason };
}

function failBadScheme(scheme: string): UrlParseFailure {
  const pretty = scheme.length > 0 ? scheme : "(none)";
  return fail(
    `Connection URL must use the postgresql scheme (postgresql, postgres or pgsql); got '${pretty}'.`,
  );
}

/** `port` as an int, or null when it is not a usable 1..65535 value. */
function parsePort(raw: string | null | undefined): number | null {
  if (raw === null || raw === undefined) return null;
  const text = raw.trim();
  if (text.length === 0) return null;
  if (!/^\d+$/.test(text)) return null;
  const value = Number(text);
  if (value < 1 || value > 65535) return null;
  return value;
}

function normalizeSslMode(value: string): string {
  const text = (value ?? "").trim();
  if (text.length === 0) return "";
  const lowered = text.toLowerCase();
  return SSL_MODES.has(lowered) ? lowered : text;
}

/** Fold the query string's sslmode into `fields` (in place). */
function applyQueryArgs(fields: Record<string, string>, query: string): void {
  if (query.length === 0) return;
  for (const pair of query.split("&")) {
    if (pair.length === 0) continue;
    const sep = pair.indexOf("=");
    const key = sep < 0 ? pair : pair.slice(0, sep);
    const value = sep < 0 ? "" : pair.slice(sep + 1);
    const name = decode(key).trim().toLowerCase();
    if (SSL_MODE_KEYS.has(name)) {
      fields.ssl_mode = decode(value).trim();
    }
  }
}

/** Percent-decode, tolerating malformed escapes (never throws). */
function decode(text: string): string {
  try {
    return decodeURIComponent(text);
  } catch {
    return text;
  }
}

export function parsePgUrl(raw: string): ParsedPgConfig | UrlParseFailure {
  if (typeof raw !== "string" || raw.trim().length === 0) {
    return fail("A connection URL is required.");
  }

  const text = raw.trim();
  if (!text.includes("://")) {
    const scheme = text.includes(":") ? (text.split(":", 1)[0] ?? "") : "";
    return failBadScheme(scheme);
  }

  const sep = text.indexOf("://");
  const scheme = text.slice(0, sep).trim().toLowerCase();
  const rest = text.slice(sep + 3);
  if (!PG_URL_SCHEMES.has(scheme)) return failBadScheme(scheme);
  if (rest.length === 0) {
    return fail("Connection URL is missing its host and database name.");
  }

  // Split off the query string before touching credentials/host.
  const q = rest.indexOf("?");
  const authorityPath = q < 0 ? rest : rest.slice(0, q);
  const query = q < 0 ? "" : rest.slice(q + 1);

  // rsplit semantics: an IPv6 host contains ':' but the userinfo '@' is the
  // real seam between credentials and host.
  const at = authorityPath.lastIndexOf("@");
  const userinfo = at < 0 ? "" : authorityPath.slice(0, at);
  const hostportPath = at < 0 ? authorityPath : authorityPath.slice(at + 1);

  const usernameRaw = userinfo.includes(":") ? userinfo.slice(0, userinfo.indexOf(":")) : userinfo;
  const passwordRaw = userinfo.includes(":") ? userinfo.slice(userinfo.indexOf(":") + 1) : "";
  const username = userinfo.length > 0 ? decode(usernameRaw).trim() : "";
  const password = passwordRaw.length > 0 ? decode(passwordRaw) : "";

  const slash = hostportPath.indexOf("/");
  const hostport = slash < 0 ? hostportPath : hostportPath.slice(0, slash);
  const path = slash < 0 ? "" : hostportPath.slice(slash);

  let host = "";
  let port = DEFAULT_PG_PORT;
  if (hostport.startsWith("[")) {
    // IPv6 literal: [::1]:5432
    const close = hostport.indexOf("]");
    if (close < 0) return fail("Connection URL has an unterminated IPv6 host literal.");
    host = hostport.slice(1, close);
    const tail = hostport.slice(close + 1);
    if (tail.length > 0) {
      const parsed = parsePort(tail.replace(/^:/, ""));
      if (parsed === null) return fail("Connection URL port must be a number between 1 and 65535.");
      port = parsed;
    }
  } else if (hostport.includes(":")) {
    const colon = hostport.indexOf(":");
    host = hostport.slice(0, colon).trim();
    const parsed = parsePort(hostport.slice(colon + 1));
    if (parsed === null) return fail("Connection URL port must be a number between 1 and 65535.");
    port = parsed;
  } else {
    host = hostport.trim();
  }

  if (host.length === 0) return fail("Connection URL is missing its host.");

  const database = decode(path.replace(/^\/+/, "")).trim();
  if (database.length === 0) return fail("Connection URL is missing its database name.");

  const fields: Record<string, string> = {
    host,
    port: String(port),
    database,
    username,
    ssl_mode: "",
  };
  applyQueryArgs(fields, query);
  const sslMode = normalizeSslMode(fields.ssl_mode ?? "");

  const cfg: ParsedPgConfig = { host, port, database, username, ssl_mode: sslMode };
  if (password.length > 0) cfg.password = password;
  return cfg;
}

export function buildPgUrl(cfg: ParsedPgConfig): string {
  const host = String(cfg.host ?? "");
  let hostPart = host;
  if (hostPart.includes(":") && !hostPart.startsWith("[")) hostPart = `[${hostPart}]`;
  const username = String(cfg.username ?? "");
  const userPart = username.length > 0 ? `${username}@` : "";
  const port = Number(cfg.port) > 0 ? Number(cfg.port) : DEFAULT_PG_PORT;
  const database = String(cfg.database ?? "");
  const sslMode = normalizeSslMode(String(cfg.ssl_mode ?? ""));
  const url = `postgresql://${userPart}${hostPart}:${port}/${database}`;
  return sslMode.length > 0 ? `${url}?sslmode=${sslMode}` : url;
}

export function maskUrl(raw: string): string {
  if (typeof raw !== "string" || !raw.includes("://")) return raw;
  const sep = raw.indexOf("://");
  const scheme = raw.slice(0, sep);
  const rest = raw.slice(sep + 3);
  const at = rest.lastIndexOf("@");
  if (at < 0) return raw;
  const creds = rest.slice(0, at);
  const tail = rest.slice(at + 1);
  const colon = creds.indexOf(":");
  if (colon < 0) return raw;
  const user = creds.slice(0, colon);
  return `${scheme}://${user}:***@${tail}`;
}
